File: Leetcode/Javascript/3. Longest Substring Without Repeating Characters.js
var lengthOfLongestSubstring = function(s) {
    var chars = new Array(256);
    for(var i = 0; i < 256; ++i) chars.push(false);
    //--------------------------------------------//
    var j = 0;
    var max = 0;
    for(i = 0; i < s.length; ++i) {
        var index = s.charCodeAt(i);
        if(!chars[index]) chars[index] = true;
        else {
            max = Math.max(i-j, max);
            var tmp = s.charCodeAt(j);
            while(tmp != index) {
                chars[tmp] = false;
                j++;
                tmp = s.charCodeAt(j);
            }
            j++;
        }
    }
    return Math.max(i-j, max);
};

File: Leetcode/Three/3. Longest Substring Without Repeating Characters.java
/*
    Hash
*/
public class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left = 0;
        Set<Character> set = new HashSet<>();
        int max = 0;
        for(int i = 0; i < s.length(); ++i) {
            char x = s.charAt(i);
            if(!set.contains(x)) {
                set.add(x);
            }
            else {
                max = Math.max(i-left, max);
                while(s.charAt(left) != x) {
                    set.remove(s.charAt(left++));
                }
                left++;
            }
        }
        return Math.max(s.length()-left, max);
    }
}
